src/RA_StringUtils.h

```diff
@@ -25,65 +25,486 @@ _NODISCARD std::string Narrow(_In_ const std::string& wstr);
+    {
+        std::ostringstream oss;
+
+        if (sFormat.back() == 'f' || sFormat.back() == 'F')
         {
-            sFormatted.resize(ra::to_unsigned(nNeeded + 1));
-            int nCheck = 0;
-            va_start(pArgs, sFormat);
-            if constexpr (std::is_same_v<CharT, char>)
-                nCheck = std::vsnprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
-            else if constexpr (std::is_same_v<CharT, wchar_t>)
-                nCheck = std::vswprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
-            va_end(pArgs);
-            assert(nCheck == nNeeded);
+            int nIndex = sFormat.find('.');
+            if (nIndex != std::string::npos)
+            {
+                int nPrecision = atoi(sFormat.c_str() + nIndex + 1);
+                oss.precision(nPrecision);
+                oss << std::fixed;
+            }
+        }
+
+        oss << arg;
+        Append(oss.str());
+    }
+
+    template<>
+    void AppendFormat(int arg, const std::string& sFormat)
+    {
+        std::ostringstream oss;
+        if (sFormat.front() == '0')
+            oss << std::setfill('0');
+        int nDigits = atoi(sFormat.c_str());
+        if (nDigits > 0)
+            oss << std::setw(nDigits);
+
+        if (sFormat.back() == 'X')
+            oss << std::uppercase << std::hex;
+        else if (sFormat.back() == 'x')
+            oss << std::hex;
+
+        oss << arg;
+        Append(oss.str());
+    }
+
+    template<>
+    void AppendFormat(unsigned int arg, const std::string& sFormat)
+    {
+        std::ostringstream oss;
+        if (sFormat.front() == '0')
+            oss << std::setfill('0');
+        int nDigits = atoi(sFormat.c_str());
+        if (nDigits > 0)
+            oss << std::setw(nDigits);
+
+        if (sFormat.back() == 'X')
+            oss << std::uppercase << std::hex;
+        else if (sFormat.back() == 'x')
+            oss << std::hex;
+
+        oss << arg;
+        Append(oss.str());
+    }
+
+    template<typename CharT, typename = std::enable_if_t<is_char_v<CharT>>>
+    void AppendPrintf(const CharT* pFormat)
+    {
+        while (*pFormat)
+        {
+            const CharT* pScan = pFormat;
+            while (*pScan && *pScan != '%')
+                ++pScan;
+
+            if (*pScan == '%')
+            {
+                ++pScan;
+                AppendSubString(pFormat, pScan - pFormat);
+
+                if (*pScan == '%')
+                    ++pScan;
+                else if (*pScan != '\0')
+                    assert(!"not enough parameters provided");
+            }
+            else if (pScan > pFormat)
+            {
+                AppendSubString(pFormat, pScan - pFormat);
+            }
+
+            pFormat = pScan;
+        }
+    }
+
+    template<typename CharT, typename = std::enable_if_t<is_char_v<CharT>>, typename T, typename... Ts>
+    void AppendPrintf(const CharT* pFormat, T value, Ts... args)
```

Comment:
If T is guaranteed to be a literal type then leave it. args should be passed by `rvalue reference` since parameter packs are forwarded. Each arg will be forwarded either as an lvalue or rvalue reference depending on the what current `Ts` is.

This is `std::forward` definitions
```
template< class T >
constexpr T&& forward( typename std::remove_reference<T>::type& t ) noexcept;

template< class T >
constexpr T&& forward( typename std::remove_reference<T>::type&& t ) noexcept;
```

`std::move` is basically the same except that's when the type is already known, `std::forward` is for templates.

Reply:
changed `T` to const reference, added && to Ts